### ISSUE TITLE
[QUESTION]: tracking training loss per target instead of average

### ISSUE THREAD

**Issue by soulios** [chemprop/chemprop#361]
Hello,

In the train.py file you have the code to track the average training_loss. I would like to track the loss per target and I am a bit stuck.
If I put an inner loop like this:
```
  
    for label_idx in range(args.num_tasks):
            mask = torch.tensor(mask_batch[label_idx], dtype=torch.bool)
            targets = torch.tensor([[0 if x is None else x for x in tb] for tb in target_batch])[:, label_idx]
            targets = targets.reshape(-1, 1)
           if args.target_weights is not None:
                target_weights = torch.tensor(args.target_weights).unsqueeze(0) # shape(1,tasks)
            else:
                target_weights = torch.ones(targets.shape[0])

            # target_weights = torch.ones(targets.shape[1]).unsqueeze(0)
            data_weights = torch.tensor(data_weights_batch).unsqueeze(1)  # shape(batch,1)
            ............
            preds = model(
                mol_batch,
                features_batch,
                atom_descriptors_batch,
                atom_features_batch,
                bond_features_batch,
            )
            preds_label = preds[:, label_idx].reshape(-1,1)
            ..........
            #take the last loss as an example
            loss = loss_func(preds_label, targets) * target_weights * data_weights 
            label_loss_sums[label_idx] += (loss * mask).sum().item() / mask.sum()

```
The training loss per target skyrockets.
Does anyone have any idea why?
Also, I am not completely sure that the inner loop is the way to go.
I am a newbie so excuse my ignorance.

**Comment by kevingreenman** [comment]
A few questions to help me understand:
1. Where exactly (line number) are you trying to insert this inner loop in [train.py](https://github.com/chemprop/chemprop/blob/master/chemprop/train/train.py)?
2. Are you trying to track the loss per target, but still take the next step based on the average? i.e. what is the `loss` you use to call `loss.backward()`?
3. When you say the loss per target skyrockets, skrockets relative to what? Relative to the average loss you were getting before?
4. What `loss_func` are you using?
5. It looks from your code like you're making identical predictions `args.num_tasks` times since your `preds = model(...)` is inside the loop, and then you're extracting the label from the model for `label_idx` and calculating the loss. Is there a reason you're doing it this way instead of doing the predictions and loss function once outside the loop and then extracting the loss for each `label_idx` inside the loop?

**Comment by soulios** [comment]
1. Line 53 before mask
2. the loss.backward is done per target I think based on this line
`            loss = loss_func(preds_label, targets) * target_weights * data_weights 
`
3.Yeah it 100x higher than the average loss before.
4.binary cross entropy (line 27 in the loss_functions.py : nn.BCEWithLogitsLoss(reduction="none")
5.My idea was to calculate the preds resulting in a matrix (batch_size, args.num_tasks), calculate the loss for all and then do the backward bsed only on the label_idx. I now realise it is really inefficient. Maybe the loss per target is really high because in a certain batch there could be only a few samples per target. But I cannot find another way to do that. Maybe create data_loaders per target? I don't know, I am a bit lost.

**Comment by kevingreenman** [comment]
Is `loss.backward()` inside your loop, or is the entirety of the loop shown in the snippet you provided above? Is your loop replacing lines 53-?, or you've inserted your loop there and then lines 54-end are still there?

What is your goal here, and why does that require the loss per target?

**Comment by soulios** [comment]
Since in most datasets there are multiple tasks, I do not get why when choosing _classification_ and not _multiclass_ there shouldn't be multiple models and each one should solve one task based on its own per target loss. This way I think you will inevitably converge to local minima in some tasks or maybe not even minima. Below I am providing the full new version of my loop below:
```
   for t in range(args.num_tasks):
        indices = [i for i, target in enumerate(data_loader.targets) if target[t] != None ]
        task = Subset(data_loader.dataset,indices)
        task_data_loader = MoleculeDataLoader(task)
        for batch in tqdm(task_data_loader , total=len(task_data_loader), leave=False):
            # Prepare batch
            batch: MoleculeDataset
            mol_batch, features_batch, target_batch, mask_batch, atom_descriptors_batch, atom_features_batch, bond_features_batch, data_weights_batch = \
                batch.batch_graph(), batch.features(), batch.targets(), batch.mask(), batch.atom_descriptors(), \
                batch.atom_features(), batch.bond_features(), batch.data_weights()
            #mask = torch.tensor(mask_batch[:, label_idx], dtype=torch.bool)
            mask = torch.tensor(mask_batch[t], dtype=torch.bool)
            targets = torch.tensor([[0 if x is None else x for x in tb] for tb in target_batch])[:, t]
            targets = targets.reshape(-1, 1)

            if args.target_weights is not None:
                target_weights = torch.tensor(args.target_weights).unsqueeze(0) # shape(1,tasks)
            else:
                target_weights = torch.ones(targets.shape[0])
            data_weights = torch.tensor(data_weights_batch).unsqueeze(1)  # shape(batch,1)

            # Run model
            model.zero_grad()
            preds = model(
                mol_batch,
                features_batch,
                atom_descriptors_batch,
                atom_features_batch,
                bond_features_batch,
            )    
            preds_label = preds[:, t].reshape(-1,1)
            # Move tensors to correct device
            torch_device = preds.device
            mask = mask.to(torch_device)
            targets = targets.to(torch_device)
            target_weights = target_weights.to(torch_device)
            data_weights = data_weights.to(torch_device)

            loss = loss_func(preds_label, targets) * target_weights * data_weights
            loss = loss.sum() / mask.sum()
            loss_sum += loss.item()     
            iter_count += 1
            loss.backward()

            if args.grad_clip:
                nn.utils.clip_grad_norm_(model.parameters(), args.grad_clip)

            optimizer.step()

            if isinstance(scheduler, NoamLR):
                scheduler.step()

            n_iter += len(batch)

            # Log and/or add to tensorboard
            if (n_iter // args.batch_size) % args.log_frequency == 0:
                lrs = scheduler.get_lr()
                pnorm = compute_pnorm(model)
                gnorm = compute_gnorm(model)
                loss_avg = loss_sum / iter_count
                # loss_sum = iter_count = 0
                loss_plot.append(loss_avg)
                lrs_str = ", ".join(f"lr_{i} = {lr:.4e}" for i, lr in enumerate(lrs))
                    for i, lr in enumerate(lrs):
                        wandb.log({"lr": lr})
                if writer is not None:
                    writer.add_scalar("train_loss", loss_avg, n_iter)
                    writer.add_scalar("param_norm", pnorm, n_iter)
                    writer.add_scalar("gradient_norm", gnorm, n_iter)
                    for i, lr in enumerate(lrs):
                        writer.add_scalar(f"learning_rate_{i}", lr, n_iter)

    return n_iter, loss_plot

```

**Comment by kevingreenman** [comment]
From what I have seen, the standard practice in deep learning is to use a combination of domain knowledge and empirical testing to determine whether a single or multiple models should be used for a multitask dataset. In cases where certain tasks are highly correlated or known from domain expertise to be related, a single model that learns the tasks together often performs better than multiple models trained on each task independently. In other cases where tasks are less correlated/related, the opposite may be true (as you point out, the local minimum reached by a single model might be worse than if one chooses to use independent models).

Is your goal with this approach to train independent models for each task in your dataset? If so, you could make separate input files for each task and train one model per input file. This wouldn't require any modifications to the source code. 

Thanks for sharing the new version of your code. From reading it, it looks like you might be trying to train a model on task 1, train the same model on task 2 (starting from the endpoint of the training on model 1), train the same model on task 3 (starting from the endpoint of the training on model 2), etc. Sort of like pretraining / transfer learning, but for different tasks in the same dataset. I don't see anywhere that the model is reinitialized, so that's why I interpret it that way. It also seems that your backpropagation / optimization steps are based only on the loss for the current task rather than on some combination of the losses for different tasks. If this is your goal (rather than independent models for each task), I think it would be easier to use Chemprop's [pretraining](https://github.com/chemprop/chemprop#pretraining) functionality to accomplish this rather than modifying the source code.

If neither of these is your goal, could you explain your goal in more detail to help me understand?

**Comment by soulios** [comment]
Thanks for your help. My goal is to train different model for each task so to create different input files would be the optimal way. Thanks for not letting me waste more of my time :smile: .
